src/Confluent.Kafka.SchemaRegistry/Serialization/AvroSerializer.cs

```diff
@@ -0,0 +1,150 @@
+
+        // Deserializers against different versions of the schema (older or newer)
+        private readonly Dictionary<int, DatumReader<object>> _avroDeserializerBySchemaId = new Dictionary<int, DatumReader<object>>();
+
+
+        /// <summary>
+        ///     Initiliaze an avro serializer.
+        /// </summary>
+        /// <param name="schemaRegistryClient"></param>
+        /// <param name="isKey"></param>
+        public AvroSerializer(ISchemaRegistryClient schemaRegistryClient, bool isKey)
+        {
+            SchemaRegistryClient = schemaRegistryClient;
+            IsKey = isKey;
+        }
+
+        private Avro.Schema FromType(Avro.Schema.Type type)
+            => Avro.Schema.Parse(Avro.Schema.GetTypeString(type));
+
+        private Avro.Schema GetSchema(object data)
+        {
+            if (data == null )
+                return FromType(Avro.Schema.Type.Null);
+            if (data is bool)
+                return FromType(Avro.Schema.Type.Boolean);
+            if (data is int)
+                return FromType(Avro.Schema.Type.Int);
+            if (data is long)
+                return FromType(Avro.Schema.Type.Long);
+            if (data is float)
+                return FromType(Avro.Schema.Type.Float);
+            if (data is double)
+                return FromType(Avro.Schema.Type.Double);
+            if (data is string)
+                return FromType(Avro.Schema.Type.String);
+            if (data is byte[])
+                return FromType(Avro.Schema.Type.Bytes);
+            if (data is ISpecificRecord specificRecord)
+                return specificRecord.Schema;
+            if (data is GenericRecord genericRecord)
+                return genericRecord.Schema;
+            
+            throw new ArgumentException(
+                    "Unsupported Avro type. Supported types are null, Boolean, Integer, Long, "
+                    + "Float, Double, String, byte[], SpecificRecord and GenericRecord");
+        }
+
+    
+        /// <summary>
+        /// Serialize data
+        /// If SchemaId is not initialized, may throw SchemaRegistry related exception
+        /// </summary>
+        /// <param name="data"></param>
+        /// <param name="topic"></param>
+        /// <param name="length">Length of the result to take into account </param>
+        /// <param name="isKey"></param>
+        /// <returns></returns>
+        public byte[] Serialize(string topic, object data)
+        {
+            string subject = SchemaRegistryClient.GetRegistrySubject(topic, IsKey);
+            // TODO check to use something else than 30 which is not optimal.
+            // For primitive type, we can "easily" generate an exact value
+            using (var stream = new MemoryStream(30))
+            {
+                int schemaId;
+                Avro.Schema schema = GetSchema(data);
+                schemaId = SchemaRegistryClient.RegisterAsync(subject, schema.ToString()).Result;
```

Comment:
We probably should have made the `Serialize` method on `ISerializable` return a `Task` and made this method `async`.

Related: if this were the case, some of the incorrectly named `ProduceAsync` methods would then actually be correctly named (the ones on the serializing producer).

Reply:
In practice this will be non-blocking the vast majority of the time, so I'm thinking we should just leave things as is. To justify changing the interface, I think we'd want to be aware of some scenarios where this would actually really matter. I can't think of any currently.

Opinions welcome.

Reply:
If we make the serializer non async, the Produce order may not be fullfilled. I don't like mixing async and non async though, or blocking on Produce, so no good opinion on this, other than "you should call register before making call to produce". On this end we could also only expose a "RetrieveId" on schemaRegistry which would retrieve the id synchronously from cache and throw if Register was not call before, but this is not user friendly...